flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveCompactReader.java

```diff
@@ -0,0 +1,147 @@
+		super(reader);
+	}
+
+	public static CompactReader.Factory<RowData> factory(
+			StorageDescriptor sd,
+			Properties properties,
+			JobConf jobConf,
+			CatalogTable catalogTable,
+			String hiveVersion,
+			RowType producedRowType,
+			boolean useMapRedReader) {
+		return new Factory(
+				sd,
+				properties,
+				new JobConfWrapper(jobConf),
+				catalogTable.getPartitionKeys(),
+				catalogTable.getSchema().getFieldNames(),
+				catalogTable.getSchema().getFieldDataTypes(),
+				hiveVersion,
+				producedRowType,
+				useMapRedReader);
+	}
+
+	/**
+	 * Factory to create {@link HiveCompactReader}.
+	 */
+	private static class Factory implements CompactReader.Factory<RowData> {
+
+		private static final long serialVersionUID = 1L;
+
+		private final StorageDescriptor sd;
+		private final Properties properties;
+		private final JobConfWrapper jobConfWrapper;
+		private final List<String> partitionKeys;
+		private final String[] fieldNames;
+		private final DataType[] fieldTypes;
+		private final String hiveVersion;
+		private final HiveShim shim;
+		private final RowType producedRowType;
+		private final boolean useMapRedReader;
+
+		private Factory(
+				StorageDescriptor sd,
+				Properties properties,
+				JobConfWrapper jobConfWrapper,
+				List<String> partitionKeys,
+				String[] fieldNames,
+				DataType[] fieldTypes,
+				String hiveVersion,
+				RowType producedRowType,
+				boolean useMapRedReader) {
+			this.sd = sd;
+			this.properties = properties;
+			this.jobConfWrapper = jobConfWrapper;
+			this.partitionKeys = partitionKeys;
+			this.fieldNames = fieldNames;
+			this.fieldTypes = fieldTypes;
+			this.hiveVersion = hiveVersion;
+			this.shim = HiveShimLoader.loadHiveShim(hiveVersion);
+			this.producedRowType = producedRowType;
+			this.useMapRedReader = useMapRedReader;
+		}
+
+		@Override
+		public CompactReader<RowData> create(CompactContext context) throws IOException {
+			HiveSourceSplit split = createSplit(context.getPath(), context.getFileSystem());
+			HiveBulkFormatAdapter format = new HiveBulkFormatAdapter(
+					jobConfWrapper, partitionKeys, fieldNames, fieldTypes, hiveVersion, producedRowType, useMapRedReader);
+			BulkFormat.Reader<RowData> reader = format.createReader(context.getConfig(), split);
+			return new HiveCompactReader(reader);
+		}
+
+		private HiveSourceSplit createSplit(Path path, FileSystem fs) throws IOException {
+			long len = fs.getFileStatus(path).getLen();
+			return new HiveSourceSplit("id", path, 0, len, new String[0], null, createPartition(path));
+		}
+
+		private HiveTablePartition createPartition(Path path) {
+			Map<String, Object> partitionSpec = new LinkedHashMap<>();
+			Map<String, DataType> nameToTypes = new HashMap<>();
+			for (int i = 0; i < fieldNames.length; i++) {
+				nameToTypes.put(fieldNames[i], fieldTypes[i]);
+			}
+			for (Map.Entry<String, String> entry : extractPartitionSpecFromPath(path).entrySet()) {
```

Comment:
Is it possible that partition spec cannot be extracted from path? E.g. when we write to a static partition and the partition location is different from the default one?

Reply:
We don't support this, now the path must be come from `table.getSd()`, and the rule must be partitioned path.

Reply:
Let's then log a jira to track this. Although not popular, it's still a valid use case in hive.

Reply:
https://issues.apache.org/jira/browse/FLINK-20040